clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -441,23 +441,34 @@ class radix_sort {
       unsigned_keys[i] = detail::traits<T>::twiddle_in(unsigned_keys[i]);
     }
 
-    while (true) {
+    for (int i = begin_bit; i < end_bit; i += RADIX_BITS) {
       int pass_bits = sycl::min(RADIX_BITS, end_bit - begin_bit);
 
       int ranks[VALUES_PER_THREAD];
       detail::radix_rank<RADIX_BITS, DESCENDING>(_local_memory)
-          .template rank_keys(item, unsigned_keys, ranks, begin_bit, pass_bits);
-      begin_bit += RADIX_BITS;
+          .template rank_keys(item, unsigned_keys, ranks, i, pass_bits);
 
       item.barrier(sycl::access::fence_space::local_space);
 
-      exchange<T, VALUES_PER_THREAD>(_local_memory)
-          .scatter_to_blocked(item, keys, ranks);
+      if (i == end_bit - RADIX_BITS) {
 
-      item.barrier(sycl::access::fence_space::local_space);
+        if (is_striped) {
+          exchange<T, VALUES_PER_THREAD>(_local_memory)
+              .scatter_to_striped(item, keys, ranks);
+
+          item.barrier(sycl::access::fence_space::local_space);
+        } else {
+          exchange<T, VALUES_PER_THREAD>(_local_memory)
+              .scatter_to_blocked(item, keys, ranks);
+
+          item.barrier(sycl::access::fence_space::local_space);
+        }
+      } else {
+        exchange<T, VALUES_PER_THREAD>(_local_memory)
+            .scatter_to_blocked(item, keys, ranks);
 
-      if (begin_bit >= end_bit)
-        break;
+        item.barrier(sycl::access::fence_space::local_space);
+      }
```

Comment:
There are a couple things still wrong here,

- I don't believe we are guaranteed that `end-bit - begin_bit` is evenly divided by `RADIX_BITS`, so lets not assume it here.
- We have the same code in both else blocks, we can just change the logic and combine those.

We are trying to merge today so here is my suggestion (please test/format):

``` 
      bool last_iter = i + RADIX_BITS > end_bit;
      if ( last_iter && is_striped ) {
          exchange<T, VALUES_PER_THREAD>(_local_memory)
            .scatter_to_striped(item, keys, ranks);
      }
      else{
          exchange<T, VALUES_PER_THREAD>(_local_memory)
            .scatter_to_blocked(item, keys, ranks);
      } 
      item.barrier(sycl::access::fence_space::local_space); 
```